File: cerebras/gtest/test_elu.cpp
// test_elu.cpp
#include <gtest/gtest.h>
#include <torch/torch.h>
#include "elu.h"

namespace custom_namespace {

    // Test for the compute_elu method with positive alpha
    TEST(EluTest, PositiveAlpha) {
        custom_namespace::ELU elu_instance;
        torch::Tensor input = torch::tensor({-1.0, 2.0, 1.0, 20.0});
        double alpha = 1.0;

        // Call compute_elu
        torch::Tensor result = elu_instance.compute_elu(input, alpha);

        // Expected output for ELU with alpha = 1.0 on [-1, 0, 1, 2]
        torch::Tensor expected = torch::elu(input, alpha);

        // Test if the result is close to the expected output
        ASSERT_TRUE(torch::allclose(result, expected, 1e-4));
    }

    // Test for the compute_elu method with a negative input tensor
    TEST(EluTest, NegativeInputs) {
        custom_namespace::ELU elu_instance;
        torch::Tensor input = torch::tensor({-1.5, -0.5, -0.1, 0.0});
        double alpha = 1.0;

        // Call compute_elu
        torch::Tensor result = elu_instance.compute_elu(input, alpha);

        // Expected output for ELU with alpha = 1.0 on negative inputs
        torch::Tensor expected = torch::elu(input, alpha);

        // Test if the result matches the expected output
        ASSERT_TRUE(torch::allclose(result, expected, 1e-4));
    }

}  // namespace custom_namespace

File: test/test_elu.py
import torch, pytest
import custom_module


@pytest.mark.parametrize(
    "input_shapes",
    (
        (torch.Size([1, 1, 32, 32])),
        (torch.Size([1, 1, 320, 384])),
        (torch.Size([1, 3, 320, 384])),
    ),
)
@pytest.mark.parametrize("alpha", [-1.0, 0.0, 1.0, 3.0])
def test_elu(input_shapes, alpha):

    # Generate Randn Input
    input = torch.randn(input_shapes).bfloat16()

    # Create an instance of Heaviside
    ch = custom_module.ELU()

    # Perform custom Heaviside
    custom_elu = ch.compute_elu(input, alpha)

    torch_elu = torch.nn.functional.elu(input, alpha)

    # Assert that the custom result matches the expected output
    assert torch.allclose(
        custom_elu, torch_elu, rtol=1e-02, atol=1e-02
    ), f"custom_elu({input}, {alpha}) != torch.elu({input}, {alpha})"
